extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java

```diff
@@ -367,19 +397,26 @@ private void createFrameWriterIfNeeded(RowsAndColumns rac, AtomicInteger rowId)
   public void writeRacToFrame(RowsAndColumns rac, AtomicInteger rowId) throws IOException
   {
     final int numRows = rac.numRows();
-    rowId.set(0);
     while (rowId.get() < numRows) {
-      final boolean didAddToFrame = frameWriter.addSelection();
-      if (didAddToFrame) {
+      if (frameWriter.addSelection()) {
+        incrementBoostColumn();
         rowId.incrementAndGet();
-        partitionBoostVirtualColumn.setValue(partitionBoostVirtualColumn.getValue() + 1);
-      } else if (frameWriter.getNumRows() == 0) {
-        throw new FrameRowTooLargeException(currentAllocatorCapacity);
-      } else {
+      } else if (frameWriter.getNumRows() > 0) {
         flushFrameWriter();
-        return;
+        createFrameWriterIfNeeded(rac, rowId);
+
+        if (frameWriter.addSelection()) {
+          incrementBoostColumn();
+          rowId.incrementAndGet();
+          return;
```

Comment:
the old code was doing kinda the same;   instead of duplicating - can we go back to that?

Reply:
What part is duplicated? The `frameWriter.addSelection()` needs to be "tried" again if it wasn't successful in the first attempt because of reaching capacity.

I could create a separate method to bundle in the incrementing of rowId and boost column, if that's what you meant.

> the old code was doing kinda the same

It looks similar, but it wasn't really doing the same thing 😅

Reply:
This was initially inspired by [GroupByPostShuffleFrameProcessor](https://github.com/apache/druid/blob/3d56fa6f5607d9b56f6da7705f4fb2ec8b6803ce/extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/groupby/GroupByPostShuffleFrameProcessor.java#L247), but later on it also helps us in taking a decision whether we need to re-run `runIncrementally()` or not. (as we can check if some row is pending flush)

Reply:
> writing a single row to use it as a sign that there is still data sounds like a shady contract

We want to know if frame writer has rows pending flush, so we are checking frame writer's current state for it 😅 - what contract would you suggest instead?